File: src/modules/students/application/student.service.ts
import { prisma } from "@/lib/db/prisma";
import type { OnboardingInput } from "./student.schemas";

export async function upsertStudentProfile(userId: string, input: OnboardingInput) {
  const budget = input.annualBudgetMinor ? BigInt(input.annualBudgetMinor) : null;
  return prisma.$transaction(async (tx) => {
    const profile = await tx.studentProfile.upsert({
      where: { userId },
      create: {
        userId,
        dateOfBirth: input.dateOfBirth ? new Date(input.dateOfBirth) : null,
        gender: input.gender,
        countryCode: input.countryCode,
        state: input.state,
        city: input.city,
        educationStage: input.educationStage,
        preferredCountries: input.preferredCountries,
        preferredLanguages: input.preferredLanguages,
        budgetCurrency: input.budgetCurrency,
        annualBudgetMinor: budget,
        onboardingCompletedAt: new Date(),
      },
      update: {
        dateOfBirth: input.dateOfBirth ? new Date(input.dateOfBirth) : null,
        gender: input.gender,
        countryCode: input.countryCode,
        state: input.state,
        city: input.city,
        educationStage: input.educationStage,
        preferredCountries: input.preferredCountries,
        preferredLanguages: input.preferredLanguages,
        budgetCurrency: input.budgetCurrency,
        annualBudgetMinor: budget,
        onboardingCompletedAt: new Date(),
      },
    });
    await Promise.all([
      tx.studentInterest.deleteMany({ where: { studentId: profile.id } }),
      tx.studentSkill.deleteMany({ where: { studentId: profile.id } }),
      tx.careerGoal.deleteMany({ where: { studentId: profile.id } }),
    ]);
    if (input.interests.length) await tx.studentInterest.createMany({ data: input.interests.map((x) => ({ studentId: profile.id, ...x })) });
    if (input.skills.length) await tx.studentSkill.createMany({ data: input.skills.map((x) => ({ studentId: profile.id, ...x })) });
    if (input.goals.length) await tx.careerGoal.createMany({ data: input.goals.map((x) => ({ studentId: profile.id, ...x })) });
    return profile;
  });
}

File: backend/app/services/student_service.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.student import CareerGoal, StudentInterest, StudentProfile, StudentSkill
from app.schemas.student import StudentProfileUpsert


async def upsert_student_profile(
    db: AsyncSession, user_id: str, payload: StudentProfileUpsert
) -> StudentProfile:
    result = await db.execute(select(StudentProfile).where(StudentProfile.user_id == user_id))
    profile = result.scalar_one_or_none()

    if profile is None:
        profile = StudentProfile(
            user_id=user_id,
            date_of_birth=payload.date_of_birth,
            gender=payload.gender,
            country_code=payload.country_code,
            state=payload.state,
            city=payload.city,
            education_stage=payload.education_stage,
            preferred_countries=payload.preferred_countries,
            preferred_languages=payload.preferred_languages,
            budget_currency=payload.budget_currency,
            annual_budget_minor=payload.annual_budget_minor,
        )
        db.add(profile)
        await db.flush()
    else:
        for key, value in payload.model_dump(exclude={"interests", "skills", "goals"}).items():
            setattr(profile, key, value)

        await db.execute(StudentInterest.__table__.delete().where(StudentInterest.student_id == profile.id))
        await db.execute(StudentSkill.__table__.delete().where(StudentSkill.student_id == profile.id))
        await db.execute(CareerGoal.__table__.delete().where(CareerGoal.student_id == profile.id))

    db.add_all([
        *(StudentInterest(student_id=profile.id, **item.model_dump()) for item in payload.interests),
        *(StudentSkill(student_id=profile.id, **item.model_dump()) for item in payload.skills),
        *(CareerGoal(student_id=profile.id, **item.model_dump()) for item in payload.goals),
    ])
    await db.commit()
    await db.refresh(profile)
    return profile
